pkg/querier/worker/scheduler_processor_test.go

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package worker
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/httpgrpc"
+	"go.uber.org/atomic"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/grafana/mimir/pkg/scheduler/schedulerpb"
+)
+
+func TestSchedulerProcessor_processQueriesOnSingleStream(t *testing.T) {
+	t.Run("should immediately return if worker context is canceled and there's no inflight query", func(t *testing.T) {
+		sp, loopClient, requestHandler := prepareSchedulerProcessor()
+
+		loopClient.On("Recv").Return(func() (*schedulerpb.SchedulerToQuerier, error) {
+			// No query to execute, so wait until terminated.
+			<-loopClient.Context().Done()
+			return nil, loopClient.Context().Err()
+		})
+
+		requestHandler.On("Handle", mock.Anything, mock.Anything).Return(&httpgrpc.HTTPResponse{}, nil)
+
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+		workerCancel()
+
+		startTime := time.Now()
+		sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		assert.Less(t, time.Since(startTime), time.Second)
```

Comment:
Nit: instead of relying on timing in the test, could the `processQueriesOnSingleStream` method return some value that could be checked instead?

Reply:
I removed this assertion because it was superfluous. I keep the one in the other test, but it can't be flaky due to timing there.